kong/clustering/services/sync/init.lua

```diff
@@ -0,0 +1,69 @@
+local _M = {}
+local _MT = { __index = _M, }
+
+
+local events = require("kong.clustering.events")
+local strategy = require("kong.clustering.services.sync.strategies.postgres")
+local rpc = require("kong.clustering.services.sync.rpc")
+
+
+-- TODO: what is the proper value?
+local FIRST_SYNC_DELAY = 0.5  -- seconds
```

Comment:
I would think we would want to sync as soon as possible, no?

```suggestion
local FIRST_SYNC_DELAY = 0
```

Reply:
In our test suits we may need to wait cp to start, here 0.5 second perhaps is a proper number, or else the next sync will be 30 seconds later.

Reply:
Ah, interesting. To me, this seems like this ought to be solved at the RPC level with events/callbacks:

1. RPC subsystem (client/driver) establishes a connection as soon as possible (nginx `init_worker`/`ngx.timer.at(0, ...)`) with a retry/exponential back-off policy
  a. Attempts to send an RPC call before the connection is established yield an error 
2. On establishing a successful connection, the RPC subsystem emits a worker-level event\*

During nginx `init`, incremental sync subsystem can register a callback for this "on-connect" event and call `:sync_once()` in addition to registering the recurring sync calls every `EACH_SYNC_DELAY` seconds. This way we also execute sync ASAP when a connection is lost and re-established. If we add a similar "on-disconnect" event, incremental sync subsystem can use it to cancel the recurring sync timer.


\* "worker-level event" doesn't _necessarily_ imply that we need to use `kong.worker_events` to drive this behavior. It could be a simple callback mechanism whereby `kong.rpc:on_connect("handler-name", callback)` adds the callback function to a table, and the RPC subsystem executes those functions at the appropriate time.

Reply:
This should be solved with callback mechanisms on RPC. It is a TODO for the technical preview.

Reply:
KAG-5569